### PR TITLE
fix: implement test.each, handle read-only props, and reset boot recovery for harness tests

### PR DESCRIPTION
**Pull request by huhuanming**
## Summary
- Implement `test.each` support in the harness jest-compat layer (1D and 2D array tables)
- Handle read-only property descriptors (Metro's `export *` getter-only re-exports) with `safeDelete`/`safeSet` helpers instead of direct `Object.assign`/`delete`
- Add `globalSetup` to reset the boot recovery counter before harness tests, preventing the crash-recovery page from blocking React Native startup during test runs (iOS via UserDefaults flag, Android via marker file)

## Test plan
- [ ] Run harness tests with `test.each` cases and verify each table entry runs as a separate test
- [ ] Verify mocking modules with read-only re-exported properties no longer throws
- [ ] Run harness tests multiple times in succession and confirm the recovery page does not appear

---

<a href="https://app.devin.ai/review/onekeyhq/app-monorepo/pull/10787" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open with Devin
  </picture>
</a>

### CODE REVIEW

**Comment by revan-zhang** [comment]
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scan Engine |  Critical |  High |  Medium |  Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/rayston92/pr-checks/d29afe32-6845-4e53-aeb4-00c6c181463c?refs=76e405ba2286dd0653a0385b097bad1c743a23a5&source=prChecksComment) |
| :white_check_mark: | Licenses| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/rayston92/pr-checks/d29afe32-6845-4e53-aeb4-00c6c181463c/license?refs=76e405ba2286dd0653a0385b097bad1c743a23a5&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

File: apps/mobile/babel-plugin-jest-compat.js [hunk 1]
```diff
@@ -52,15 +52,42 @@ module.exports = function ({ types: t }) {
...
+                  t.blockStatement([
+                    t.returnStatement(
+                      t.callExpression(t.identifier('require'), [args[0]]),
+                    ),
+                  ]),
+                  t.catchClause(
+                    t.identifier('_e'),
+                    t.blockStatement([
+                      t.returnStatement(t.objectExpression([])),
+                    ]),
```
[the remark this answers is not in the document]
Comment (huhuanming) [reply apps/mobile/babel-plugin-jest-compat.js:93]: Fixed: Added console.warn to the generated catch block so require() failures are logged instead of silently swallowed. The try-catch still returns {} to prevent crashes from native module init failures, but the warning makes resolution/init errors visible for debugging.

File: apps/mobile/jest.harness.config.mjs [hunk 2]
```diff
@@ -1,6 +1,7 @@
 /** @type {import('jest').Config} */
 export default {
   preset: 'react-native-harness',
+  globalSetup: '<rootDir>/harness/globalSetup.mjs',
```
[the remark this answers is not in the document]
Comment (huhuanming) [reply apps/mobile/jest.harness.config.mjs:4]: Fixed: Added globalTeardown (harness/globalTeardown.mjs) that clears the harness_mode flag on Android (rm marker file) and iOS (defaults delete) after the test session ends, preventing stale flags on reused emulators/simulators.

File: apps/mobile/harness/globalSetup.mjs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    }
+  }
+}
+
+export default async function globalSetup() {
+  // Try both platforms — one will succeed, the other will silently fail.
+  try {
+    setAndroidHarnessFlag();
+  } catch {
+    // No Android device/emulator connected — expected on iOS runs
```
[the remark this answers is not in the document]
Comment (huhuanming) [reply apps/mobile/harness/globalSetup.mjs:86]: Fixed: Now tracking success of each platform setup and logging the error message when either fails. If both platforms fail, a console.warn is emitted to make the setup failure visible.

File: apps/mobile/harness/jest-compat.ts [hunk 4]
```diff
@@ -106,11 +145,65 @@ type ModSnapshot = {
...
+
 const saveSnapshot = (mod: Record<string, unknown>) => {
   if (mockSnapshots.has(mod)) return;
   const snapshot: ModSnapshot = { top: {} };
   for (const key of Object.keys(mod)) {
-    snapshot.top[key] = mod[key];
+    try {
+      snapshot.top[key] = mod[key];
+    } catch {
+      // Getter threw — skip this key in snapshot
```
[the remark this answers is not in the document]
Comment (huhuanming) [reply apps/mobile/harness/jest-compat.ts]: Fixed: Using a GETTER_THREW sentinel symbol for keys whose getter throws during snapshot. restoreAllMocks now skips these keys instead of treating them as missing (which would incorrectly delete the export).

**Review by sidmorizon — DISMISSED** [review]

File: apps/mobile/harness/globalSetup.mjs [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    setAndroidHarnessFlag();
+    androidOk = true;
+  } catch (e) {
+    // No Android device/emulator connected — expected on iOS runs
+    console.log(`[harness-globalSetup] Android flag skipped: ${e.message}`);
+  }
+  try {
+    setIOSHarnessFlag();
+    iosOk = true;
+  } catch (e) {
```
[the remark this answers is not in the document]
Comment (huhuanming) [reply apps/mobile/harness/globalSetup.mjs]: Fixed: setIOSHarnessFlag now returns a boolean indicating whether at least one booted simulator was updated. iosOk is only set to true when a flag was actually written.

File: apps/mobile/harness/jest-compat.ts [hunk 6]
```diff
@@ -224,9 +339,13 @@ const restoreAllMocks = () => {
       // Mutate the module exports directly
       const keys = Object.keys(mod).filter((k) => k !== '__esModule');
       for (const key of keys) {
-        delete mod[key];
+        safeDelete(mod, key);
+      }
+      for (const key of Object.keys(mockExports)) {
+        if (key !== '__esModule') {
+          safeSet(mod, key, mockExports[key]);
+        }
```
[the remark this answers is not in the document]
Comment (huhuanming) [reply apps/mobile/harness/jest-compat.ts:364]: Fixed: All safeSet call sites in __harness_mock_module__ now check the return value and emit a console.warn when a key cannot be mutated, making silent mock failures visible.

File: apps/mobile/android/app/src/main/java/so/onekey/app/wallet/MainApplication.java [hunk 7]
```diff
@@ -107,7 +107,10 @@ public void onCreate() {
     int newCount = oldCount + 1;
     prefs.edit().putInt(BootRecoveryKeys.CONSECUTIVE_BOOT_FAIL_COUNT, newCount).commit();
 
-    shouldShowRecovery = newCount >= 3;
+    // Harness tests create this marker file via globalSetup so the recovery
+    // page never blocks React Native from starting during test runs.
+    boolean isHarnessMode = new java.io.File(getFilesDir(), "harness_mode").exists();
+    shouldShowRecovery = !isHarnessMode && newCount >= 3;
```
[the remark this answers is not in the document]
Comment (huhuanming) [reply apps/mobile/android/app/src/main/java/so/onekey/app/wallet/MainApplication.java:113]: Fixed: globalTeardown now also resets the boot-fail counter on both platforms — Android (sed on SharedPrefs XML) and iOS (defaults delete on onekey_consecutive_boot_fail_count) — so harness restarts do not trigger recovery on the next normal launch.

File: apps/mobile/harness/jest-compat.ts [hunk 8]
```diff
@@ -198,9 +308,15 @@ const restoreAllMocks = () => {
         if (!(mockExports as any).__esModule && !mockExports.default) {
           const keys = Object.keys(defaultObj);
           for (const key of keys) {
-            delete defaultObj[key];
+            safeDelete(defaultObj, key);
+          }
```
[the remark this answers is not in the document]
Comment (huhuanming) [reply apps/mobile/harness/jest-compat.ts:316]: Fixed: All safeDelete call sites in __harness_mock_module__ now check the return value and emit a console.warn when an export cannot be removed, making partial mocks visible.

File: apps/mobile/harness/globalTeardown.mjs [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+  );
+  const { devices } = JSON.parse(out);
+  for (const runtime of Object.values(devices)) {
+    for (const device of runtime) {
+      if (device.state === 'Booted') {
+        execFileSync(
+          'xcrun',
+          [
+            'simctl',
+            'spawn',
```
[the remark this answers is not in the document]
Comment (huhuanming) [reply apps/mobile/harness/globalTeardown.mjs]: Fixed: Wrapped each per-device defaults delete in its own try/catch so a failure on one simulator does not abort cleanup of remaining devices.

File: apps/mobile/harness/globalSetup.mjs [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+  try {
+    iosOk = setIOSHarnessFlag();
+    if (!iosOk) {
+      console.log('[harness-globalSetup] iOS: no booted simulators found');
+    }
+  } catch (e) {
+    // No booted iOS simulator — expected on Android runs
+    console.log(`[harness-globalSetup] iOS flag skipped: ${e.message}`);
+  }
+  if (!androidOk && !iosOk) {
```
[the remark this answers is not in the document]
Comment (huhuanming) [reply apps/mobile/harness/globalSetup.mjs]: Fixed: Added a summary log line showing which platforms succeeded/skipped (e.g. "Result: Android=ok, iOS=skipped") so the outcome is always visible regardless of which platform is the actual target.

File: apps/mobile/harness/globalTeardown.mjs [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+    }
+  }
+}
+
+export default async function globalTeardown() {
+  try {
+    clearAndroidHarnessFlag();
+  } catch {
+    // No Android device/emulator — expected on iOS runs
+  }
```
[the remark this answers is not in the document]
Comment (huhuanming) [reply apps/mobile/harness/globalTeardown.mjs:111]: Fixed: Top-level catch blocks now log the error message so cleanup failures are distinguishable from the expected "platform not present" case.

File: apps/mobile/harness/globalSetup.mjs [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+  const { devices } = JSON.parse(out);
+  let flagSet = false;
+  for (const runtime of Object.values(devices)) {
+    for (const device of runtime) {
+      if (device.state === 'Booted') {
+        execFileSync(
+          'xcrun',
+          [
+            'simctl',
+            'spawn',
```
[the remark this answers is not in the document]
Comment (huhuanming) [reply apps/mobile/harness/globalSetup.mjs]: Fixed: Wrapped per-device execFileSync in try/catch so a failing simulator does not abort the loop. Failed devices are logged and the loop continues to remaining simulators.

File: apps/mobile/harness/globalSetup.mjs [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+  } catch (e) {
+    // No booted iOS simulator — expected on Android runs
+    console.log(`[harness-globalSetup] iOS flag skipped: ${e.message}`);
+  }
+  console.log(
+    `[harness-globalSetup] Result: Android=${androidOk ? 'ok' : 'skipped'}, iOS=${iosOk ? 'ok' : 'skipped'}`,
+  );
+  if (!androidOk && !iosOk) {
+    console.warn(
+      '[harness-globalSetup] WARNING: Failed to set harness flag on both platforms. ' +
```
[the remark this answers is not in the document]
Comment (huhuanming) [reply apps/mobile/harness/globalSetup.mjs]: Won't fix: There is no reliable way to detect the target platform at globalSetup time — no environment variable distinguishes iOS vs Android harness runs. The current approach already logs per-platform results and warns when both fail, which is sufficient for diagnosing setup issues.

File: apps/mobile/harness/globalTeardown.mjs [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+    // SharedPrefs file may not exist yet — not critical
+  }
+  console.log('[harness-globalTeardown] Android harness_mode flag cleared');
+}
+
+function clearIOSHarnessFlag() {
+  const out = execFileSync(
+    'xcrun',
+    ['simctl', 'list', 'devices', 'booted', '-j'],
+    {
```
[the remark this answers is not in the document]
Comment (huhuanming) [reply apps/mobile/harness/globalTeardown.mjs:44]: Won't fix: Persisting UDIDs from setup to teardown would require cross-process file-based state sharing, adding complexity for a rare edge case (simulator shutting down between setup and teardown). The current approach cleans all booted simulators at teardown time, which covers the normal case. Simulators that shut down mid-test are an unlikely scenario that doesn't justify the added complexity.

File: apps/mobile/harness/globalSetup.mjs [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+    }
+  } catch (e) {
+    // No booted iOS simulator — expected on Android runs
+    console.log(`[harness-globalSetup] iOS flag skipped: ${e.message}`);
+  }
+  console.log(
+    `[harness-globalSetup] Result: Android=${androidOk ? 'ok' : 'skipped'}, iOS=${iosOk ? 'ok' : 'skipped'}`,
+  );
+  if (!androidOk && !iosOk) {
+    console.warn(
```
[the remark this answers is not in the document]
Comment (huhuanming) [reply apps/mobile/harness/globalSetup.mjs]: Fixed: globalSetup now reads HARNESS_PLATFORM env var and throws an error if the target platform's flag write fails. The harness:test:android script passes HARNESS_PLATFORM=android so Android failures are no longer masked by an iOS success.

File: apps/mobile/harness/globalTeardown.mjs [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+    const key = 'consecutive_boot_fail_count';
+    const sedCmd = `sed -i 's/${key}" value="[0-9]*/${key}" value="0/' shared_prefs/onekey_recovery.xml`;
+    execFileSync(
+      'adb',
+      ['shell', 'run-as', 'so.onekey.app.wallet', 'sh', '-c', sedCmd],
+      { stdio: 'pipe', timeout: 5000 },
+    );
+  } catch {
+    // SharedPrefs file may not exist yet — not critical
+  }
```
[the remark this answers is not in the document]
Comment (huhuanming) [reply apps/mobile/harness/globalTeardown.mjs]: Fixed: The empty catch block now logs the failure message so Android counter-reset issues are visible in test output.

File: apps/mobile/harness/globalTeardown.mjs [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+              'defaults',
+              'delete',
+              'so.onekey.wallet',
+              'onekey_harness_mode',
+            ],
+            { stdio: 'pipe', timeout: 5000 },
+          );
+        } catch {
+          // Key may not exist on this simulator — continue to next device
+        }
```
[the remark this answers is not in the document]
Comment (huhuanming) [reply apps/mobile/harness/globalTeardown.mjs]: Fixed: Both iOS catch blocks now log the specific error message per device and track cleanup status. The final log line reports "partial" instead of "done" when any deletion failed.

File: apps/mobile/harness/globalSetup.mjs [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+    throw new Error(
+      '[harness-globalSetup] Android is the target runner but flag setup failed',
+    );
+  }
+  if (targetPlatform === 'ios' && !iosOk) {
+    throw new Error(
+      '[harness-globalSetup] iOS is the target runner but flag setup failed',
+    );
+  }
+  if (!targetPlatform && !androidOk && !iosOk) {
```
[the remark this answers is not in the document]
Comment (huhuanming) [reply apps/mobile/harness/globalSetup.mjs:114]: Fixed: Added HARNESS_PLATFORM=ios to the default harness:test script to match the defaultRunner config in rn-harness.config.mjs. Both scripts now explicitly declare their target platform.

**Review by originalix — DISMISSED** [review]

File: apps/mobile/harness/jest-compat.ts [hunk 19]
```diff
@@ -267,10 +403,51 @@ Object.defineProperty(globalThis, 'jest', {
...
     },
+    // Fake timers cannot be safely implemented in the harness because
+    // replacing globalThis.setTimeout breaks the harness bridge communication.
+    // Tests using fake timers are excluded via jest.harness.config.mjs.
+    // These no-op stubs prevent crashes if a test slips through.
+    useFakeTimers: () => {
+      console.warn(
+        '[harness-compat] jest.useFakeTimers() is not supported — test excluded from harness?',
+      );
+      return (globalThis as any).jest;
```
[the remark this answers is not in the document]
Comment (huhuanming) [reply apps/mobile/harness/jest-compat.ts]: Fixed: useFakeTimers now throws an error with a clear message pointing to testPathIgnorePatterns, so tests that slip through fail explicitly instead of silently running with real timers.

File: apps/mobile/babel-plugin-jest-compat.js [hunk 20]
```diff
@@ -52,15 +52,64 @@ module.exports = function ({ types: t }) {
...
+                          [
+                            t.stringLiteral(
+                              '[babel-plugin-jest-compat] require() failed for mock:',
+                            ),
+                            t.logicalExpression(
+                              '||',
+                              t.memberExpression(
+                                t.identifier('_e'),
+                                t.identifier('message'),
+                              ),
```
[the remark this answers is not in the document]
Comment (huhuanming) [reply apps/mobile/babel-plugin-jest-compat.js]: Fixed: Replaced `_e.message || _e` with `String(_e)` in the generated catch block, which safely handles null/undefined/non-Error throws without risking a secondary exception.

**Review by originalix — APPROVED** [review]

**Review by limichange — APPROVED** [review]
